### PR TITLE
draft-ietf-ippm-ioam-data-integrity-09

### PR DESCRIPTION
**Pull request by IurmanJ**
draft-ietf-ippm-ioam-data-integrity-09

### CODE REVIEW

File: drafts/draft-ietf-ippm-ioam-data-integrity.xml [hunk 1]
```diff
@@ -524,312 +500,327 @@
...
+          participate (in that order) in the integrity protection of
+          IOAM-Data-Fields:
+            <list style="numbers">
+              <t>Namespace-ID</t>
+              <t>IOAM-E2E-Type</t>
+            </list>
+
+          Those are all the header fields defined for a E2E Option-Type.</t>
+
+          <t>A document defining a new IOAM Integrity Protected Option-Type that
```
Comment (fbrockne) [inline drafts/draft-ietf-ippm-ioam-data-integrity.xml:746]: s/A document defining a new IOAM Integrity Protected Option-Type that intends to/New IOAM Integrity Protected Option-Types that intend to/

File: drafts/draft-ietf-ippm-ioam-data-integrity.xml [hunk 2]
```diff
@@ -524,312 +500,327 @@
...
         </t>
-      </list>
-      </t>
 
-      <t>The method assumes that symmetric keys have been distributed to
-      the respective nodes as well as the Validator(s). The details of the
-      mechanisms used for key distribution are outside the scope of this
-      document.</t>
+        <t>The transit node updates the ICV field in the IOAM Integrity
+        Protection header. Reminder: it MUST always be a 16-octet ICV.</t>
```
Comment (fbrockne) [inline drafts/draft-ietf-ippm-ioam-data-integrity.xml:777]: Remove: "Reminder: it MUST always be a 16-octet ICV." - see above.

File: drafts/draft-ietf-ippm-ioam-data-integrity.xml [hunk 3]
```diff
@@ -524,312 +500,327 @@
...
+
+        <t>If the transit node does not add any IOAM-Data-Fields (e.g., it
+        only modifies mutable IOAM-Data-Fields or does nothing), then the
+        transit node MUST NOT update the ICV field in the IOAM Integrity
+        Protection header.</t>
```
Comment (fbrockne) [inline drafts/draft-ietf-ippm-ioam-data-integrity.xml:782]: Is this really true? In case of pre-allocated trace, the fields would be *updated* and not added, but you still want the ICV field to be updated, wouldn't you?

File: drafts/draft-ietf-ippm-ioam-data-integrity.xml [hunk 4]
```diff
@@ -524,312 +500,327 @@
...
+      </section>
+
+      <section anchor="IPM-Decap" title="Decapsulating node">
+        <t>In the context of integrity protection, the decapsulating node can
+        either act as a transit node (see <xref target="IPM-Transit"/> above) or
+        directly as the Validator (see <xref target="IPM-Validator"/> below).
+        The former is an alternative to put the Validator out of the forwarding
```
Comment (fbrockne) [inline drafts/draft-ietf-ippm-ioam-data-integrity.xml:789]: IMHO this could be confusing. You make a difference between "IOAM transit nodes" and "IOAM-IPM transit nodes".
Just calling a decap node a transit node could give raise to mis-interpretations. IMHO we should either use "IOAM transit nodes, IOAM decap nodes" and "IOAM-IPM transit nodes" etc. throughout the doc or re-phrase this paragraph.

E.g., you could say: The Validator role can either be performed by the IOAM decapsulating node, or another node.

Question: If the Validator is not the IOAM decap node, how would this work, given that the decap node removes the IOAM header?

File: drafts/draft-ietf-ippm-ioam-data-integrity.xml [hunk 5]
```diff
@@ -524,312 +500,327 @@
...
+      </section>
+
+      <section anchor="IPM-Validator" title="Validator">
+        <t>A node that performs the validation of the integrity protection is
+        referred to as the Validator. This method assumes that symmetric keys
+        have been distributed to the Validator only. The details of the
+        mechanisms used for key distribution are outside the scope of this
+        document. We refer to the method as "Zero Trust" in the sense that
+        neither the encapsulating node, nor transit nodes, nor any other
+        non-IOAM nodes need to be trusted. Indeed, the Validator is the only
```
Comment (fbrockne) [inline drafts/draft-ietf-ippm-ioam-data-integrity.xml:800]: s/Indeed, the/The/

File: drafts/draft-ietf-ippm-ioam-data-integrity.xml [hunk 6]
```diff
@@ -524,312 +500,327 @@
...
+        previous steps of the method in the same order, using the respective
+        symmetric keys received previously. The recomputed ICV is then compared
+        to the received ICV field. As a result, the Validator can detect if the
+        integrity of IOAM-Data-Fields is intact or altered. The validation is
+        one-step in some cases (e.g., with POT Type-0 or E2E), where only the
+        encapsulating node updates the ICV, according to the definition of this
+        method. For other cases where transit nodes also update the ICV (e.g.,
+        with Trace Option-Types), the Validator MUST identify these transit
+        nodes in order to look up their respective keys. For that, something
+        unique representing a node MUST be included in IOAM-Data-Fields (e.g.,
```
Comment (fbrockne) [inline drafts/draft-ietf-ippm-ioam-data-integrity.xml:814]: s/For that, something unique representing a node MUST be included in IOAM-Data-Fields (e.g., the "node_id" for Trace Option-Types)./For that, a unique identifier of the node, such as the "node_id" for Trace Option-Types MUST be included in IOAM-Data-Fields/

File: drafts/draft-ietf-ippm-ioam-data-integrity.xml [hunk 7]
```diff
@@ -524,312 +500,327 @@
...
+        the "node_id" for Trace Option-Types). Details on how the mapping
+        between those identifiers and keys is implemented on the Validator are
+        outside the scope of this document.</t>
+
+        <t>The Validator MUST NOT update the ICV field in the IOAM Integrity
+        Protection header. Since its role is to validate the integrity of
+        IOAM-Data-Fields, it trusts itself whether it adds IOAM-Data-Fields or
+        not.</t>
```
Comment (fbrockne) [inline drafts/draft-ietf-ippm-ioam-data-integrity.xml:822]: This is only feasible if the Validator is the decap node, correct?

Are we also dealing with cases where only a part of the overall IOAM "journey" is integrity protected?
IMHO we should exclude this in the requirements part, i.e. for IOAM integrity protection the entire header is protected and an IOAM transit node can neither add nor remove the IOAM integrity protection header,

File: drafts/draft-ietf-ippm-ioam-data-integrity.xml [hunk 8]
```diff
@@ -912,18 +894,59 @@
...
-      prevent replay attacks. To enable replay protection, the encapsulating node
-      and the Validator(s) MUST agree on a common methodology to keep the Nonce valid
-      only for a specific period of time, which is outside the scope of this document.</t>
+      <t>There is an additional per-packet processing for each node that uses
+      the Integrity Protection Method defined in this document. Inappropriate
+      use of this Integrity Protection Method might overload nodes and cause
+      them to stop functioning properly. Operators deploying IOAM with this
+      Integrity Protection Method MUST ensure that such overload situations are
+      avoided. This could for example be achieved by applying IOAM only to a
+      subset of the entire traffic.</t>
```
Comment (fbrockne) [inline drafts/draft-ietf-ippm-ioam-data-integrity.xml:903]: Mention the obvious caveat: If you don't protect all packets with IOAM, the unprotected packets are unprotected.

File: drafts/draft-ietf-ippm-ioam-data-integrity.xml [hunk 9]
```diff
@@ -912,18 +894,59 @@
...
+      nonce (AES-GMAC IV) and a key MUST NOT be used more than once. Otherwise,
+      security guarantees are destroyed. To avoid such scenario, and to avoid
+      frequent rotation or refreshing of keys, it is RECOMMENDED to use a
+      12-octet (i.e., 96-bit) nonce based on the "Deterministic Construction"
+      (see <xref target="NIST.800-38D"/>, Sec. 8). It is defined as follows:
+        <list style="bullets">
+          <t>The nonce is the concatenation of two fields, called the fixed
+          field and the invocation field.</t>
+
+          <t>The fixed field identifies the device and MUST be unique. It can be
```
Comment (fbrockne) [inline drafts/draft-ietf-ippm-ioam-data-integrity.xml:918]: Is there a potential to re-use the node_id?
Node_id short is 3 bytes (which could fit into the 32 bits needed here), Node_id wide is 7 bytes.

**Review by fbrockne — CHANGES_REQUESTED** [review]
Hey Justin, 
thanks a lot for crafting the new rev. Overall it looks very good and reflects the plan outlined and discussed with the WG. I caught a few nits while reviewing.
Cheers, Frank

File: drafts/draft-ietf-ippm-ioam-data-integrity.xml [hunk 10]
```diff
@@ -524,312 +500,327 @@
...
-        <list style="bullets">
-          <t>If the transit node does not insert IOAM-Data-Fields
-          (e.g., it only modifies IOAM-Data-Fields in-place, or does nothing),
-          then the transit node MUST NOT generate a signature and MUST NOT
-          update the Signature field.</t>
-        </list>
+      <t>This document defines a new method that uses a symmetric key based
+      algorithm for the integrity protection of IOAM-Data-Fields. In case of
+      performance concerns, the method can be applied to a subset of the traffic
+      by using sampling of data.</t>
```
[the remark this answers is not in the document]
Comment (IurmanJ) [reply drafts/draft-ietf-ippm-ioam-data-integrity.xml:659]: I'd suggest to remove the second sentence (i.e., "In case of performance concerns, ..."), since we already discuss it in the Security section. But still, your remark makes sense and I'll add a sentence in the Security section. Sounds good?

File: drafts/draft-ietf-ippm-ioam-data-integrity.xml [hunk 11]
```diff
@@ -524,312 +500,327 @@
...
-        specifies the length of the Nonce in octets.</t>
-
-        <t hangText="Reserved:">16-bit Reserved field. MUST be set to zero
+        <t hangText="Method-ID:">8-bit unsigned integer.
+        It defines the Integrity Protection Method to compute the Integrity
+        Check Value (ICV) field. If a node encounters an unknown value, it MUST
+        NOT change the contents of the IOAM Integrity Protection header and MUST
+        NOT change the contents of the IOAM-Data-Fields. In other words, the
+        node MUST NOT process the IOAM Option-Type.
+        See <xref target="IOAM-IP-registry"/>.</t>
```
[the remark this answers is not in the document]
Comment (IurmanJ) [reply drafts/draft-ietf-ippm-ioam-data-integrity.xml:541]: Indeed. I think it's expected. If a node does not implement an Integrity Protection Method (identified by the "Method-ID" field), then it is not able to update the ICV. Should it add its IOAM-Data-Fields no matter what, the Validator would consider the whole thing as altered. So I think it's probably better to force a node not to add its data in that case. Should I add a sentence to reflect your remark? Thoughts on the above?

File: drafts/draft-ietf-ippm-ioam-data-integrity.xml (hunk 3 above)
Comment (IurmanJ) [reply drafts/draft-ietf-ippm-ioam-data-integrity.xml:782]: Indeed. Do you have a better wording in mind?

File: drafts/draft-ietf-ippm-ioam-data-integrity.xml (hunk 11 above)
[the remark this answers is not in the document]
Comment (fbrockne) [reply drafts/draft-ietf-ippm-ioam-data-integrity.xml:541]: Hmmm... Assume you have a node that cannot do ICV but it does support IOAM. It could update the IOAM-data.
Then it is up to the validator to decide whether the info would still be used or not.
IMHO we don't need to be overly restrictive.

File: drafts/draft-ietf-ippm-ioam-data-integrity.xml (hunk 3 above)
Comment (fbrockne) [reply drafts/draft-ietf-ippm-ioam-data-integrity.xml:782]: How about:

"If the transit node does not add or change any IOAM-Data-Fields (e.g., it only modifies mutable IOAM-Data-Fields or does nothing), then the transit node MUST NOT update the ICV field in the IOAM Integrity Protection header."

Thinking about things again: Why do we even deal with this special case? There would be no harm for the node to update the ICV field - even if everything is unchanged. Plus you would know that the node was indeed visited (kind of "proof of transit" through a backdoor).

File: drafts/draft-ietf-ippm-ioam-data-integrity.xml (hunk 9 above)
Comment (IurmanJ) [reply drafts/draft-ietf-ippm-ioam-data-integrity.xml:918]: Good catch, thanks! It was indeed the plan, so that the Validator can easily identify the encapsulating node and use the right key. I'll add it to the recommendations.

Comment (IurmanJ) [reply drafts/draft-ietf-ippm-ioam-data-integrity.xml:918]: Should we force the recommendations for the nonce construction too, i.e., s/RECOMMENDED/MUST/g ? Obviously for interop reasons, but no strong opinion though.

Comment (fbrockne) [reply drafts/draft-ietf-ippm-ioam-data-integrity.xml:918]: MUST limits options - which is a good thing in this case

Comment (IurmanJ) [reply drafts/draft-ietf-ippm-ioam-data-integrity.xml:918]: > MUST limits options - which is a good thing in this case

Not sure what part you referred to here?

File: drafts/draft-ietf-ippm-ioam-data-integrity.xml (hunk 3 above)
Comment (IurmanJ) [reply drafts/draft-ietf-ippm-ioam-data-integrity.xml:782]: > How about:
> 
> "If the transit node does not add or change any IOAM-Data-Fields (e.g., it only modifies mutable IOAM-Data-Fields or does nothing), then the transit node MUST NOT update the ICV field in the IOAM Integrity Protection header."

"or change" could be problematic (see below).

> Thinking about things again: Why do we even deal with this special case? There would be no harm for the node to update the ICV field - even if everything is unchanged. Plus you would know that the node was indeed visited (kind of "proof of transit" through a backdoor).

Think about an Option-Type with a mutable data field (currently, you only have the "Cumulative" field in POT type 0, but who knows what will come next). If a node modifies the value and updates the ICV, the integrity validation will always fail (there is no way to get previous values of the field). On the other hand, if everything is unchanged, it means the node cannot be identified in the data fields and so MUST NOT update the ICV, otherwise the integrity validation will always fail too (in addition to being a total waste of cpu cycles).

How about:

"If the transit node only modifies mutable IOAM-Data-Fields or does nothing, then the transit node MUST NOT update the ICV field in the IOAM Integrity Protection header."

In other words, is it better by removing the first part, i.e., "If the transit node does not add any IOAM-Data-Fields" ?

File: drafts/draft-ietf-ippm-ioam-data-integrity.xml (hunk 9 above)
Comment (fbrockne) [reply drafts/draft-ietf-ippm-ioam-data-integrity.xml:918]: I was referring to your statement; "Should we force the recommendations for the nonce construction too, i.e., s/RECOMMENDED/MUST/g ? Obviously for interop reasons, but no strong opinion though."

IMHO changing to MUST is reasonable.

Comment (IurmanJ) [reply drafts/draft-ietf-ippm-ioam-data-integrity.xml:918]: Got it! Sorry, I thought it was some proposed text :-D

File: drafts/draft-ietf-ippm-ioam-data-integrity.xml (hunk 11 above)
[the remark this answers is not in the document]
Comment (IurmanJ) [reply drafts/draft-ietf-ippm-ioam-data-integrity.xml:541]: What I fear is to have either (i) interop issues or (ii) an easy kill for compromised nodes. The integrity validation will never work in both cases (e.g., a new method is implemented by some of your nodes, not all; or malicious behaviors on compromised nodes). Of course the Validator can do what it wants with IOAM-Data-Fields, but what's the most probable outcome for an "invalid integrity" signal...? Which would be equivalent to "no IOAM at all for the node" IMHO. But I definitely understand why your opinion makes sense too. I guess it all comes down to whether we still want IOAM to "work" as expected, regardless the integrity validation is broken.

File: drafts/draft-ietf-ippm-ioam-data-integrity.xml (hunk 3 above)
Comment (IurmanJ) [reply drafts/draft-ietf-ippm-ioam-data-integrity.xml:782]: Correct. The confusion comes from "mutable vs immutable fields" for Trace Option-Types. With Pre-allocated, it could indeed be considered mutable fields since it is first pre-allocated and then "modified" by the node. With Incremental, it can only be considered immutable fields. TBH, I don't know how to rephrase this to make it work :-D

Comment (fbrockne) [reply drafts/draft-ietf-ippm-ioam-data-integrity.xml:782]: Given that the mechanism does not really work for fields like "cumulative", could we just limit ICV to trace and E2E option types?

File: drafts/draft-ietf-ippm-ioam-data-integrity.xml (hunk 4 above)
Comment (IurmanJ) [reply drafts/draft-ietf-ippm-ioam-data-integrity.xml:789]: > IMHO this could be confusing. You make a difference between "IOAM transit nodes" and "IOAM-IPM transit nodes". Just calling a decap node a transit node could give raise to mis-interpretations. IMHO we should either use "IOAM transit nodes, IOAM decap nodes" and "IOAM-IPM transit nodes" etc. throughout the doc or re-phrase this paragraph.

+1 to re-phrase the paragraph (I prefer that vs the other solution).

> E.g., you could say: The Validator role can either be performed by the IOAM decapsulating node, or another node.

Not sure how to keep a sub-section for the decapsulating node in that case. I'll think about it tonight and come with proposed text tomorrow.

> Question: If the Validator is not the IOAM decap node, how would this work, given that the decap node removes the IOAM header?

Most obvious idea I have in mind: a packet copy is sent by the decapsulating node (before the IOAM removal) to the Validator. There are other possibilities, but I think we should stay away from this. Maybe add a sentence to say that the decapsulating node would need to send IOAM data to the Validator, but is out of scope?

Comment (fbrockne) [reply drafts/draft-ietf-ippm-ioam-data-integrity.xml:789]: > Maybe add a sentence to say that the decapsulating node would need to send IOAM data to the Validator, but is out of scope?

That could work if you're really keen to keep that option. We could say something like - "If the decapsulating node does not perform the function of the Validator, the decapsulating node needs to send the IOAM related fields to the Validator. The method how the IOAM data is sent to the Validator is out of scope for this document".

File: drafts/draft-ietf-ippm-ioam-data-integrity.xml (hunk 7 above)
Comment (IurmanJ) [reply drafts/draft-ietf-ippm-ioam-data-integrity.xml:822]: > This is only feasible if the Validator is the decap node, correct?

Well, yes but it could also be "another" node (that's the reason why the proposed text was kind of generic).

> Are we also dealing with cases where only a part of the overall IOAM "journey" is integrity protected?

Not sure I understand your question.

> IMHO we should exclude this in the requirements part, i.e. for IOAM integrity protection the entire header is protected

I probably missed something above, which confuses me here too. What do you mean by "the entire header is protected"? Some header fields MUST NOT be included, otherwise there is no way to recompute the ICV. Also, exclude what from (what) requirements?

> and an IOAM transit node can neither add nor remove the IOAM integrity protection header,

+1

Comment (fbrockne) [reply drafts/draft-ietf-ippm-ioam-data-integrity.xml:822]: What I mean with "part of the journey" is the following:
Consider a network A -- B -- C --- D --- E.

Only the part B -- C -- D is across some "dangerous" territory,

Could I protect just the journey from B to C to D with the ICV?

File: drafts/draft-ietf-ippm-ioam-data-integrity.xml (hunk 4 above)
Comment (IurmanJ) [reply drafts/draft-ietf-ippm-ioam-data-integrity.xml:789]: +1, thanks! Indeed, I think it'd be good to have both alternatives, in case decap==Validator is too heavy to deploy depending on situations.

File: drafts/draft-ietf-ippm-ioam-data-integrity.xml (hunk 7 above)
Comment (IurmanJ) [reply drafts/draft-ietf-ippm-ioam-data-integrity.xml:822]: Got it. Indeed, we should exclude this, as a protected segment is always from the encap to the decap, not a subset.

File: drafts/draft-ietf-ippm-ioam-data-integrity.xml (hunk 9 above)
Comment (IurmanJ) [reply drafts/draft-ietf-ippm-ioam-data-integrity.xml:918]: I suggest to s/RECOMMENDED/MUST/g and get feedback from the WG on this. WDYT?

File: drafts/draft-ietf-ippm-ioam-data-integrity.xml (hunk 11 above)
[the remark this answers is not in the document]
Comment (IurmanJ) [reply drafts/draft-ietf-ippm-ioam-data-integrity.xml:541]: I suggest to keep it as is for -09 and ask feedback from the WG on whether they think A or B is more appropriate. WDYT?

File: drafts/draft-ietf-ippm-ioam-data-integrity.xml (hunk 9 above)
Comment (fbrockne) [reply drafts/draft-ietf-ippm-ioam-data-integrity.xml:918]: sounds good

File: drafts/draft-ietf-ippm-ioam-data-integrity.xml (hunk 3 above)
Comment (fbrockne) [reply drafts/draft-ietf-ippm-ioam-data-integrity.xml:782]: How about we remove it for now - and then ask the WG for either acknowledging - or proposing a solution which would work.

File: drafts/draft-ietf-ippm-ioam-data-integrity.xml (hunk 11 above)
[the remark this answers is not in the document]
Comment (fbrockne) [reply drafts/draft-ietf-ippm-ioam-data-integrity.xml:541]: Works for me.

File: drafts/draft-ietf-ippm-ioam-data-integrity.xml (hunk 3 above)
Comment (IurmanJ) [reply drafts/draft-ietf-ippm-ioam-data-integrity.xml:782]: Works for me. Note that in the first paragraph describing a transit node, we also have "[...] and **immutable** IOAM-Data-Fields **added** by the transit node". Same for the encapsulating node. We'll have to determine if we keep those as is (especially considering the pre-allocated trace case).

**Review by fbrockne — COMMENTED** [review]
LGTM. Thanks for all the updates!
